test/plugin.test.js

```diff
@@ -104,31 +104,40 @@ async function publishMessage (server, topic, message, headers = {}) {
   return res
 }
 
-async function waitForCondition (conditionFn) {
-  const timeoutMs = 5000
-  const intervalMs = 100
-  const startTime = Date.now()
+function waitForMetricUpdate (registry, metricName, labelMatcher) {
+  const { promise, resolve } = Promise.withResolvers()
 
-  while (Date.now() - startTime < timeoutMs) {
-    if (await conditionFn()) {
-      return true
-    }
-    await new Promise(resolve => setTimeout(resolve, intervalMs))
-  }
-
-  throw new Error(`Condition not met within ${timeoutMs}ms`)
-}
-
-async function waitForMetricValue (registry, metricName, labelMatcher) {
-  await waitForCondition(async () => {
+  const checkMetric = async () => {
     const metric = registry.getSingleMetric(metricName)
     if (!metric) return false
 
     const metricValue = await metric.get()
     const matchingMetric = metricValue.values.find(v => labelMatcher(v.labels))
 
-    return matchingMetric && matchingMetric.value >= 1
+    if (matchingMetric && matchingMetric.value >= 1) {
+      resolve(matchingMetric)
+      return true
+    }
+    return false
+  }
+
+  checkMetric().then(resolved => {
+    if (resolved) return
+    let retry = 0
+    const intervalId = setInterval(async () => {
+      const resolved = await checkMetric()
+      if (resolved) {
+        clearInterval(intervalId)
+      }
+      if (retry > 10) {
+        clearInterval(intervalId)
+        throw new Error(`checkMetric retried ${retry} times`)
+      }
+      retry++
+    }, 200)
   })
```

Comment:
Why do you need an interval? I _think_ the metrics will be automatically added once the external message has been processed.

Reply:
An `interval` (even if very short) is needed here because metrics apparently are not immediately updated after the message is processed

Reply:
updated